src/packaging/requirements.py

```diff
@@ -56,8 +57,12 @@ def __init__(self, requirement_string: str) -> None:
             self.marker = Marker.__new__(Marker)
             self.marker._markers = _normalize_extra_values(parsed.marker)
 
-    def __str__(self) -> str:
-        parts: List[str] = [self.name]
+    @property
+    def canonical_name(self) -> str:
+        return canonicalize_name(self.name)
+
+    def _to_str(self, name: str) -> str:
```

Comment:
Can I ask what adding `_to_str` brings over directly defining `__str__`?

Reply:
Maybe it’d be a good idea to make this `_iter_parts` and simply `yield` the parts. `__str__` would do `"".join(self._iter_parts(self.name))`, and `__hash__` would `hash((..., self._iter_parts(canonicalize_name(self.name))))`.

Reply:
Got it, done.

Reply:
Yeah sorry I missed a `*`, should be `hash((..., *self._iter_parts(...)))`